active/0000-serve-file-api.md

```diff
@@ -0,0 +1,70 @@
+
+An addon can define custom logic on how it wants to serve every asset request. The public API will look as follows:
+```javascript
+  /**
+   * This hook allows you to provide a custom handler for serving assets during development using
+   * `ember *serve`. It is invoked from the `serve-files` middleware.
+   *
+   * @param {Object} request - Represents the HTTP Request object that is used in any express middleware
+   * @param {Object} response - Represents the HTTP Response object that is used in any express middleware
+   * @param {Function} next - Represents the next middleware function to run after the current.
+   * @param {Object} data - object containing per request info
+   *                [data.filename] - the current absolute path of the asset being served
+   *                [data.outputPath] - the current path where all assets are being served from
+   * @returns {null}
+   */
+  serveFileHandler: function(request, response, next, data) {
+    // serve the incoming request file with a custom logic and end the current response.
+    // For example:
+    // if (request.url === request.serveUrl) {
+    //  var fastboot = new FastBoot(data.outputPath);
+    //  fastboot.visit(request.url).then(result => result.html()).then(html => response.end(html));
+    // }
+  }
+```
+
+The above API will only be invoked during development when using `ember serve`. The above API can also be used by addons that wish to change/add response headers before an asset is served. The [serve-files](https://github.com/ember-cli/ember-cli/blob/master/lib/tasks/server/middleware/serve-files/index.js#L18) will check if any addon defines this function and will pass it down to `broccoli-middleware`. This middleware (broccoli-middleware) will then invoke the above defined function right before it is about to serve the files. It is the responsibility of this function to serve the assets. If this function is not defined, it will default to serving the file using the logic [here](https://github.com/ember-cli/broccoli-middleware/blob/master/lib/middleware.js#L109).
+
+In general we only expect one addon (representing a platform) to provide this hook to override the serve file behavior. If there is more than one addon that is providing this hook, then `ember serve` should throw an error. This is because we do not expect more than one addon to decide how the assets should be served.
+
+Having said that, two or more addons representing different infrastructures may be used together in an app. In order to allow that, we will expose an API via environment.js where an app can specify that it wants override this behavior of throwing an error if more than one addon defines `serveFileHandler` function. If this API is used, the last addon will win always.
```

Comment:
We should likely not speculate on how to solve this. Lets define the problem we want to solve, then work together to figure out how to solve.  I think this falls into "Unanswered questions"